### ISSUE TITLE
LL adds foreign language titles as new

### ISSUE THREAD

**Issue by dkids** [DobyTang/LazyLibrarian#1331]
Repo: https://github.com/dobytang/lazylibrarian : Branch: master : Updated: Mon Apr 16 14:27:04 2018
Current Version: 1b78da566f43490c4599a51c184898f9acf4b6cf : Latest Version: 1b78da566f43490c4599a51c184898f9acf4b6cf

MacOS 10.13.4

Import Languages:
[ en, eng, en-US, en-GB ]

Missing Book Status:
[ Skipped ]
Status for missing/deleted ebooks

New Book Status:
[ Wanted ]
Status for new ebooks by existing authors

What I think should happen: books with a publication year of 2018 (maybe 2017) or later are added as wanted.   Books published before then (including 0000) are added as "skipped".

What is happening: foreign language books with publication year of 1971 or 0000 are being added as wanted.   As I see them in the logs, I'm flipping them to "skipped" but it's not much fun.

![screen shot 2018-04-17 at 2 59 52 pm](https://user-images.githubusercontent.com/30055574/38901666-ecf12236-4251-11e8-994e-0cf06b84e502.png)
![screen shot 2018-04-17 at 3 02 08 pm](https://user-images.githubusercontent.com/30055574/38901667-ed3b22e6-4251-11e8-8af1-b452689a7fa3.png)

**Comment by philborman** [comment]
"New ebooks" in this context is books that are not already in our database, ie "new found" rather than "new written". I don't think it's particularly foreign language books that are affected, should be all. Goodreads data is constantly evolving (improving?) and as new books are added we try to add them to our database, but we can't tell if it's a genuinely newly written book, or one that goodreads has only just added.

Goodreads isn't always correct on the language or the dates in the details it gives us, and the details given via their api are not as comprehensive as the html author and book pages. Quite often find the html includes details that are not in the xml returned via the api. Goodreads don't like you scraping their html. It seems to be getting better over time though, fewer missing/incorrect entries.

0000 can mean it's an old book but goodreads doesn't know the date, or it's a future unspecified release date. Some future dates are just a year, some are a full date. The 1970 date is even worse, it means goodreads gave us an invalid date in the date field (or it was genuinely released in 1970 of course).

Lots of goodreads books have no language code included in their api results, so we try to look it up from other sources. We can use the isbn, as they are released in blocks per country, but that doesn't always work, publishers sometimes release foreign titles under english isbn numbers. We can try looking up the language from LibraryThing, but that sometimes gives us the "original language" the book was written in rather than the language we have got. GoogleBooks is even worse, it often assumes books are in english when it doesn't know.

Not easy to work around it

**Comment by philborman** [comment]
The first book "Het Beste boek" returns no isbn/isbn13 from goodreads, has no language code, and the title is not recognised at librarything or googlebooks so we should mark the language as "Unknown". Might be worth checking in your database and logs in case we are getting it wrong.

"Modigliani Scandal B Spl" has no dates at goodreads, and no language code, but does have both isbn and isbn13 and librarything and googlebooks both tell us that it's language is "eng".

Game of Thrones is an audiobook - we might be able to filter that out if you don't want audiobooks. It doesn't have an isbn/isbn13 at goodreads, and the title returns an english isbn from both googlebooks and librarything. 

Not sure how we can improve on this, maybe don't include anything with a missing isbn at goodreads? any suggestions?

**Comment by dkids** [comment]
You're right, it's a tough problem.

TLDR: could we set the status for everything questionable to "ignored"?   It's easy enough to search for that and then they could either be set to "skipped" or "wanted".   I think it's going to be difficult to come up with rules that are 100% effective because the info on Goodreads is not going to be perfectly correct.

I do think what they do is mostly better than my local library, where there might be 3 printings of a book and they're listed as separate items.   

Doing a weekly or monthly check on books for "ignored" isn't fun but it would be relatively easy to do for people who want to take the time to keep their libraries clean.    I started out loose (add everything as wanted) and things exploded because secondary authors would be added, then their secondary authors were added, etc.  Even now I find things attributed to "Unknown Author 812".   https://www.goodreads.com/search?q=unknown+author+812
[812.txt](https://github.com/DobyTang/LazyLibrarian/files/1929496/812.txt)


I was looking at the API for Goodreads and was concerned about where they say they may not release some information because it's not theirs but licensed from somebody else.   

"Get the reviews for a book given a Goodreads book id
Get an XML or JSON response that contains embed code for the iframe reviews widget. The reviews widget shows an excerpt (first 300 characters) of the most popular reviews of a book for a given internal Goodreads book_id. Reviews of all known editions of the book are included.

XML responses also include shelves and book meta-data (title, author, et cetera). The Goodreads API gives you full access to Goodreads-owned meta-data, but it does not give you full access to book meta-data supplied by third parties such as Ingram. Book cover images, descriptions, and other data from third party sources might be excluded, because we do not have a license to distribute these data via our API." 

Goodreads does things like add "untitled" for books (ex: [](https://www.goodreads.com/book/show/39841372-untitled); [](https://www.goodreads.com/book/show/35209671-untitled)  ;  [](https://www.goodreads.com/book/show/37842140-untitled).   I haven't worked with it enough to know what they do when the book is actually released.   Presumably it changes to 37842140-actual-title

grep -i "scandal B spl" ~/src/LazyLibrarian-master/Logs/lazylibrarian.log.* | cut -f2- -d":" | sort 
17-Apr-2018 14:39:34 - DEBUG   :: AUTHORUPDATE : gr.py:get_author_books:719 : Book found: Modigliani Scandal B Spl 0000
17-Apr-2018 14:39:36 - DEBUG   :: AUTHORUPDATE : bookwork.py:getBookCover:1272 : getBookCover: Caching goodreads cover for Ken Follett Modigliani Scandal B Spl
17-Apr-2018 14:39:36 - DEBUG   :: AUTHORUPDATE : gr.py:get_author_books:729 : Updated cover for Modigliani Scandal B Spl using goodreads
17-Apr-2018 14:39:37 - DEBUG   :: AUTHORUPDATE : gr.py:get_author_books:774 : [Ken Follett] Added book: Modigliani Scandal B Spl [eng] status Wanted
17-Apr-2018 17:41:01 - DEBUG   :: WEBSERVER : webServe.py:markBooks:2198 : Status set to "Skipped" for "Modigliani Scandal B Spl"

https://www.goodreads.com/book/show/39860976-modigliani-scandal-b-spl?ac=1&from_search=true   This book is listed as first published in 1900 and no language listed so I'm not sure where the [eng] is coming from.

I tried searching for "b spl" on Goodreads and only found 7 other books with that in the title.   So it doesn't look like it's worth the effort to block that phrase specifically.

Books are published after an author has died but they're not published before they're born.   It's not clear that there's enough of those for a check to be worth the effort.

grep -i "das lied"  ~/src/LazyLibrarian-master/Logs/lazylibrarian.log.* | grep -v zNab | grep -v torrentp | cut -f2- -d":" | sort
[das.txt](https://github.com/DobyTang/LazyLibrarian/files/1929466/das.txt)

There are results in here for [nl] and [dut].   It seems like eventually something went wrong and the book was entered as English.   I do wonder about Goodreads inquiring to other databases, getting a timeout, and returning the default of English.     "Het Beste Boke" is the "The Best Book" in Dutch.

https://www.goodreads.com/search?q=het+beste+boek has a bunch of results.

https://www.goodreads.com/book/show/26092221-het-beste-boek-1981-0621-6 is what was matched but it's also listed as Dutch.
[het.txt](https://github.com/DobyTang/LazyLibrarian/files/1929436/het.txt)

**Comment by philborman** [comment]
https://www.goodreads.com/book/show/39860976-modigliani-scandal-b-spl?ac=1&from_search=true
There is no language, but there is an isbn and an isbn13. If you look those up using thinglang
https://www.librarything.com/api/thingLang.php?isbn=1509807322
it reports [eng]   The '150' part is supposedly english, so it seems the publisher gave the book the wrong isbn?

The other book, https://www.goodreads.com/book/show/26092221-het-beste-boek-1981-0621-6
is a similar problem. Although the book page has the language as Dutch, the author page via goodreads api https://www.goodreads.com/author/list/3447.xml?page=4&key=ckvsiSDsuqh7omh74ZZ6Q doesn't give the language, and also doesn't give the isbn, so we ask google for the isbn of a book called "het+beste+boek+1981+0621+6+ken+follett" and feed that to thinglang which says [eng]

If you run "refresh active authors" with debug logging on it will tell you how many isbns we had to look up (and how long it took) so you can see how many are missing at goodreads. It's a bit of a problem.

**Comment by philborman** [comment]
Added options in config to ignore books where goodreads doesn't know the isbn, or goodreads doesn't know the publication date.  If you have the book in your library we add it anyway, but otherwise we skip it until goodreads info is updated.
